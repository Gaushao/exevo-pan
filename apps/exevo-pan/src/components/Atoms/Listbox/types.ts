import { HTMLAttributes } from 'react'
import { OptionProps } from '../Option/types'

export type ListboxProps = {
  highlightedIndex?: number
  selectedIndex?: Set<number>
  onSelectOption?: (option: Option) => void
  children: JSX.Element & React.ReactElement<OptionProps>
} & HTMLAttributes<HTMLDivElement>
