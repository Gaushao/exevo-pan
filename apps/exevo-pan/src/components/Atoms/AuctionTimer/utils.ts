import { MILLISECONDS_IN } from 'utils'
import { CountdownObject } from './types'

export const calcCountdown = (
  startTime: number,
  endTime: number,
): CountdownObject => {
  const timeDiff = endTime - startTime

  const days = Math.floor(timeDiff / MILLISECONDS_IN.DAY)
  const hours = Math.floor(
    (timeDiff % MILLISECONDS_IN.DAY) / MILLISECONDS_IN.HOUR,
  )
  const minutes = Math.floor(
    (timeDiff % MILLISECONDS_IN.HOUR) / MILLISECONDS_IN.MINUTE,
  )
  const seconds = Math.floor((timeDiff % MILLISECONDS_IN.MINUTE) / 1000)

  return { timeDiff, days, hours, minutes, seconds }
}
