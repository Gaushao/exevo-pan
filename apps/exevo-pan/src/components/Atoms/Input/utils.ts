import { Mask } from './types'

export const defaultMask: Mask = (value) => value
