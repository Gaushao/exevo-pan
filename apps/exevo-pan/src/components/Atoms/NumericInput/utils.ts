export const parseValue = (value: string) => Number(value.replace(/,/g, ''))
