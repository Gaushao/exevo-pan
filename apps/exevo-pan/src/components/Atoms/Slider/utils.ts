import { Range, TransformFunction } from './types'

export const defaultTransform: TransformFunction = (value: number) => value

export const clampValue = (value: number, [min, max]: Range): number => {
  if (value < min) return min
  if (value > max) return max
  return value
}

export const normalize = (value: number, [min, max]: Range): number =>
  (value - min) / (max - min)

export const denormalize = (value: number, [min, max]: Range): number =>
  (max - min) * value + min

const closestDistance = (a: number, b: number): number => {
  if (Math.abs(a) < Math.abs(b)) return a
  return b
}

const countDecimals = (value: number): number => {
  const [, decimals] = value.toString().split('.')

  return decimals?.length ?? 0
}

export const toFixedPrecision = (value: number, step: number): number =>
  parseFloat(value.toFixed(countDecimals(step)))

export const calculateClosestStep = (
  currentValue: number,
  step: number,
  [min, max]: Range,
): number => {
  if (step === 1) return Math.round(currentValue)
  if (currentValue === max) return max

  const adjustedValue = currentValue - min

  const distanceFromPreviousStep = (adjustedValue % step) * -1
  const distanceFromNextStep = step - (adjustedValue % step)
  const closestStepDistance = closestDistance(
    distanceFromPreviousStep,
    distanceFromNextStep,
  )

  const calculatedNewValue = adjustedValue + closestStepDistance + min

  return toFixedPrecision(calculatedNewValue, step)
}

export const getLeftOffset = (
  value: number,
  range: Range,
  invert: boolean,
): string => {
  const offset = normalize(value, range)
  return `${(invert ? 1 - offset : offset) * 100}%`
}

export const getKeyboardIncrement = (event: React.KeyboardEvent): number => {
  const sign: number = (
    {
      ArrowUp: 1,
      ArrowRight: 1,
      ArrowDown: -1,
      ArrowLeft: -1,
    } as Record<string, number>
  )[event.key]

  if (sign === undefined) return 0

  const { ctrlKey, shiftKey } = event
  return sign * (+!ctrlKey || 10) * (+!shiftKey || 100)
}

export const isInRange = (value: number, [min, max]: Range): boolean =>
  value >= min && value <= max
