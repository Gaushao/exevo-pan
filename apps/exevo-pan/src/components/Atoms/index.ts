export { default as Accordion } from './Accordion'
export { default as ActiveCount } from './ActiveCount'
export { default as AuctionTimer } from './AuctionTimer'
export { default as Button } from './Button'
export { default as Checkbox } from './Checkbox'
export { default as Chip } from './Chip'
export { default as CopyButton } from './CopyButton'
export { default as CtaButton } from './CtaButton'
export { default as Dialog } from './Dialog'
export { default as Drawer } from './Drawer'
export { default as DrawerFooter } from './DrawerFooter'
export { default as FadeImage } from './FadeImage'
export { default as FormError } from './FormError'
export { default as Input } from './Input'
export { default as Label } from './Label'
export { default as LabeledCard } from './LabeledCard'
export { default as LabeledTextBox } from './LabeledTextBox'
export { default as LazyRender } from './LazyRender'
export { default as Link } from './Link'
export { default as Listbox } from './Listbox'
export { default as LoadingAlert } from './LoadingAlert'
export { default as NumericInput } from './NumericInput'
export { default as Option } from './Option'
export { default as Paginator } from './Paginator'
export { default as PieClock } from './PieClock'
export { default as Popover } from './Popover'
export { default as RadioButton } from './RadioButton'
export { default as RadioGroup } from './RadioGroup'
export { default as RareFrame } from './RareFrame'
export { default as Skeleton } from './Skeleton'
export { default as Slider } from './Slider'
export { default as Sticker } from './Sticker'
export { default as SpritePortrait } from './SpritePortrait'
export { default as Stepper } from './Stepper'
export { default as Switch } from './Switch'
export { default as Table } from './Table'
export { default as Tabs } from './Tabs'
export { default as Tag } from './Tag'
export * as Text from './Text'
export { default as TextArea } from './TextArea'
export * as TibiaBlackjack from './TibiaBlackjack'
export { default as TimeInput } from './TimeInput'
export { default as TitledCard } from './TitledCard'
