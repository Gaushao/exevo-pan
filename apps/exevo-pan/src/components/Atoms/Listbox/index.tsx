import {
  forwardRef,
  Ref,
  useMemo,
  Children,
  isValidElement,
  cloneElement,
  memo,
} from 'react'
import clsx from 'clsx'
import { ListboxProps } from './types'
import { indexToId } from './utils'

const Listbox = (
  {
    className,
    children,
    highlightedIndex,
    selectedIndex = new Set<number>([]),
    onSelectOption,
    ...props
  }: ListboxProps,
  refProp: Ref<HTMLDivElement>,
): JSX.Element => {
  const currentActiveDescendantId = useMemo(
    () => indexToId(highlightedIndex, props.id),
    [highlightedIndex, props.id],
  )
  return (
    <div
      ref={refProp}
      role="listbox"
      aria-activedescendant={currentActiveDescendantId}
      tabIndex={-1}
      className={clsx(
        'custom-scrollbar bg-surface overflow-y-auto overflow-x-hidden rounded-md shadow-lg',
        className,
      )}
      {...props}
    >
      {Children.map(children, (child, index) => {
        if (!isValidElement(child)) return child
        if (typeof child.type === 'string') return child

        return cloneElement(child, {
          id: indexToId(index, props.id),
          highlighted: highlightedIndex === index,
          'aria-selected': selectedIndex.has(index),
          onClick: onSelectOption,
        })
      })}
    </div>
  )
}

export default memo(forwardRef(Listbox))
