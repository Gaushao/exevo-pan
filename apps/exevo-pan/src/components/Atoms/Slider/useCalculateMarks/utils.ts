import { getLeftOffset } from '../utils'
import { CalculatedMarks } from './types'
import { Range, TransformFunction } from '../types'

export const generateAllMarks = (
  step: number,
  [min, max]: Range,
  transformDisplayedValues: TransformFunction = (value): number => value,
): CalculatedMarks[] => {
  const marks: CalculatedMarks[] = []

  const pushMark = (value: number): void => {
    marks.push({
      value,
      label: transformDisplayedValues(value),
      leftOffset: getLeftOffset(value, [min, max], false),
    })
  }

  for (let currentStep = min; currentStep < max; currentStep += step) {
    pushMark(currentStep)
  }
  pushMark(max)

  return marks
}
