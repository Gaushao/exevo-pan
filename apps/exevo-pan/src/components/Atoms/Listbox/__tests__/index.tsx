import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { Option } from 'components/Atoms'
import { renderWithProviders, setup, assertNoA11yViolations } from 'utils/test'
import Listbox from '..'

const mockedOnSelectOption = jest.fn()
const mockedScrollIntoView = setup.scrollIntoView()

describe('<Listbox />', () => {
  beforeEach(() => {
    mockedScrollIntoView.mockReset()
  })

  test('should render with no highlighted <Option />', () => {
    renderWithProviders(
      <Listbox>
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )

    expect(screen.getByRole('listbox')).not.toHaveAttribute(
      'aria-activedescendant',
    )

    const optionElements = screen.queryAllByRole('option')

    optionElements.forEach((option, index) => {
      /* expect(option).toHaveStyle('background-color: #FFFFFF') */
      expect(option).toMatchSnapshot()
      expect(option).not.toHaveAttribute('aria-selected', 'true')
      expect(option).toHaveAttribute('id', `listbox-item-${index}`)
    })
  })

  test('should render with a highlighted <Option />', () => {
    const { rerender } = renderWithProviders(
      <Listbox highlightedIndex={1}>
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )

    expect(screen.getByRole('listbox')).toHaveAttribute(
      'aria-activedescendant',
      'listbox-item-1',
    )
    expect(mockedScrollIntoView).toBeCalledTimes(1)

    const [optionA, optionB, optionC] = screen.queryAllByRole('option')
    /* expect(optionA).toHaveStyle('background-color: #FFFFFF')
    expect(optionB).toHaveStyle('background-color: #C5CAE9')
    expect(optionC).toHaveStyle('background-color: #FFFFFF') */
    expect(optionA).toMatchInlineSnapshot(`
      <option
        aria-selected="false"
        class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
        id="listbox-item-0"
        value="option A"
      >
        option A
      </option>
    `)
    expect(optionB).toMatchInlineSnapshot(`
      <option
        aria-selected="false"
        class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-primaryVariant"
        id="listbox-item-1"
        value="option B"
      >
        option B
      </option>
    `)
    expect(optionC).toMatchInlineSnapshot(`
      <option
        aria-selected="false"
        class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
        id="listbox-item-2"
        value="option C"
      >
        option C
      </option>
    `)

    rerender(
      <Listbox highlightedIndex={2}>
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )
    expect(mockedScrollIntoView).toBeCalledTimes(2)

    /* expect(optionA).toHaveStyle('background-color: #FFFFFF')
    expect(optionB).toHaveStyle('background-color: #FFFFFF')
    expect(optionC).toHaveStyle('background-color: #C5CAE9') */
    expect(optionA).toMatchInlineSnapshot(`
      <option
        aria-selected="false"
        class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
        id="listbox-item-0"
        value="option A"
      >
        option A
      </option>
    `)
    expect(optionB).toMatchInlineSnapshot(`
      <option
        aria-selected="false"
        class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
        id="listbox-item-1"
        value="option B"
      >
        option B
      </option>
    `)
    expect(optionC).toMatchInlineSnapshot(`
      <option
        aria-selected="false"
        class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-primaryVariant"
        id="listbox-item-2"
        value="option C"
      >
        option C
      </option>
    `)
  })

  test('should render with a few selected', () => {
    const { rerender } = renderWithProviders(
      <Listbox selectedIndex={new Set([0, 2])}>
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )

    const [optionA, optionB, optionC] = screen.queryAllByRole('option')
    expect(optionA).toHaveAttribute('aria-selected', 'true')
    expect(optionB).toHaveAttribute('aria-selected', 'false')
    expect(optionC).toHaveAttribute('aria-selected', 'true')

    rerender(
      <Listbox selectedIndex={new Set([1])}>
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )

    expect(optionA).toHaveAttribute('aria-selected', 'false')
    expect(optionB).toHaveAttribute('aria-selected', 'true')
    expect(optionC).toHaveAttribute('aria-selected', 'false')
  })

  test('clicking an <Option /> should trigger the implicitly passed onClick prop', () => {
    renderWithProviders(
      <Listbox onSelectOption={mockedOnSelectOption}>
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )

    const [optionA, optionB, optionC] = screen.queryAllByRole('option')
    userEvent.click(optionA)
    expect(mockedOnSelectOption).toHaveBeenLastCalledWith({
      name: 'option A',
      value: 'option A',
    })

    userEvent.click(optionB)
    expect(mockedOnSelectOption).toHaveBeenLastCalledWith({
      name: 'option B',
      value: 'option B',
    })

    userEvent.click(optionC)
    expect(mockedOnSelectOption).toHaveBeenLastCalledWith({
      name: 'option C',
      value: 'option C',
    })
  })

  test('should not break with native children', () => {
    renderWithProviders(
      <Listbox onSelectOption={mockedOnSelectOption}>
        <option>option A</option>
        <option>option B</option>
        <option>option C</option>
      </Listbox>,
    )

    expect(screen.getByRole('listbox')).toBeInTheDocument()
    screen.queryAllByRole('option').forEach((option) => {
      expect(option).toBeInTheDocument()
    })
  })

  test('a11y', async () => {
    const { container } = renderWithProviders(
      <Listbox aria-label="label">
        <Option>option A</Option>
        <Option>option B</Option>
        <Option>option C</Option>
      </Listbox>,
    )
    await assertNoA11yViolations(container)
  })
})
