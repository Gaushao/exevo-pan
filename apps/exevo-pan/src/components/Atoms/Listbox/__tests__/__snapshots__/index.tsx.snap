// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Listbox /> should render with no highlighted <Option /> 1`] = `
<div
  aria-selected="false"
  class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
  data-value="option A"
  id="listbox-item-0"
  role="option"
>
  option A
</div>
`;

exports[`<Listbox /> should render with no highlighted <Option /> 2`] = `
<div
  aria-selected="false"
  class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
  data-value="option B"
  id="listbox-item-1"
  role="option"
>
  option B
</div>
`;

exports[`<Listbox /> should render with no highlighted <Option /> 3`] = `
<div
  aria-selected="false"
  class="text-tsm text-onSurface hover:bg-primaryVariant w-full cursor-pointer py-2 px-3 font-light transition-colors bg-surface"
  data-value="option C"
  id="listbox-item-2"
  role="option"
>
  option C
</div>
`;
